addons/mail/static/src/models/attachment/attachment.js

```diff
@@ -68,7 +79,7 @@ function factory(dependencies) {
             if (!attachments.includes(attachment)) {
                 return;
             }
-            this.env.messaging.dialogManager.open('mail.attachment_viewer', {
+            attachment._attachmentViewerDialog = this.env.messaging.dialogManager.open('mail.attachment_viewer', {
```

Comment:
No, attachment viewer opens attachments and one of them is active. It can change, so storing the dialog in 1st active attachment is wrong.

And that's without considering that this by-passes field, which is also wrong.

Reply:
I agree but using `dialogManager.open` on `mail.attachment_viewer` will automatically create a new `attachment_viewer` instead of using `messaging.attachmentViewer`. So we need to refactor the way `dialogManager` is doing so it allows to use another record (or none) than the one he created as it's not possible for the moment. Or directly call methods on `messaging.attachmentViewer` to show/hide himself ...

Reply:
Dialog and attachment viewer need to be redesigned in some ways, but this will have to wait. In the meantime, I would make `dialogManager.open()` support options (3rd parameter), and introduce one option to re-use a UI record.

Reply:
Still got a problem : on dialog close, the record is deleted so `env.messaging.attachmentViewer` is deleted and should be recreated after but when recreated by the dialog manager, it's not affected to `env.messaging.attachmentViewer` ... So should I change `Dialog` model to allow somehow to prevent deletion of linked record on `delete` or `DialogManager.open` so we can create a record that can be assigned to `env.messaging.attachmentViewer` ?

Reply:
Went with first option

Reply:
Hm... I think all this goes against the intended solution. Attachment viewer being a singleton is not intended, and it's there due to bad design in the 1st place. Fixing this by keeping it singleton but working around is not good.

I thought it was too complicated to break it, but now I doubt it... If it's easier to fix it by making attachment viewer no longer a singleton, then I'm all for it and would definitely pick this approach.

Sorry for turning side on this problem...

Reply:
So if I understood well, I need to delete `env.messaging.attachmentViewer` creation and allow the dialogManager to create/destroy it at each time we open/close it (and so no need anymore of the tweaking of Dialog allowing to use an existing record) ? How do you call the close on dialog now ?

Reply:
> So if I understood well, I need to delete `env.messaging.attachmentViewer` creation and allow the dialogManager to create/destroy it at each time we open/close it (and so no need anymore of the tweaking of Dialog allowing to use an existing record) ?

Yes

> How do you call the close on dialog now ?

Instead of `attachmentViewer.delete()`, it should be a `attachmentViewer.close()` that invoke `dialog.close()`.

Reply:
What do you think of that ?
I'm writing test waiting for your answer, I think this bug should have been covered by a test.

Reply:
Looks good! Yes, attachment viewer is not tested at all right now.

Reply:
Just testing here in attachment_tests that on click on attachment_tests, it shows the attachment_viewer.
Should also test in attachment_box that if two attachments you can go from one to another.
Tests of attachment viewer is just layout tests I think

Reply:
Sounds good!

Technically actions to change active attachment is on attachment viewer, so I would expect such tests being in attachment_viewer tests... But at the same time you want to test that opening of attachment viewer from some place registers all contextual attachments in viewer, and only way to check that visually is to click on next/previous buttons.

I think ideal making of tests is a test in attachment viewer to check next/previous work as expected, and a test in attachment box to consider all attachments of attachment box, etc...

But I'm totally fine with just attachment box tests for that, because it covers a lot just by itself.